crates/telio-wg/src/no_link_detection.rs

```diff
@@ -0,0 +1,524 @@
+            0,
+            0,
+            Instant::now(),
+            Instant::now(),
+            LinkState::Up,
+        );
+
+        time::advance(Duration::from_secs(1)).await;
+
+        nld.insert(&key, Some(5), Some(10), Some(LinkState::Up));
+        check_node_info(
+            nld.info.get(&key).unwrap(),
+            5,
+            10,
+            Instant::now(),
+            Instant::now(),
+            LinkState::Up,
+        );
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn test_time_since_last_rx() {
+        let Env { key, mut nld } = Env::new();
+        assert_eq!(nld.time_since_last_rx(&key), None);
+
+        // The case where it is inserted by update
+        nld.update(&key, None, None, NodeState::Connected, false);
+        assert_eq!(nld.time_since_last_rx(&key), Some(Instant::now().elapsed()));
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.insert(&key, None, None, None);
+        assert_eq!(nld.time_since_last_rx(&key), Some(Instant::now().elapsed()));
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.insert(&key, None, None, Some(LinkState::Up));
+        assert_eq!(nld.time_since_last_rx(&key), Some(Instant::now().elapsed()));
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.insert(&key, Some(5), Some(5), Some(LinkState::Up));
+        assert_eq!(nld.time_since_last_rx(&key), Some(Instant::now().elapsed()));
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.update(&key, Some(10), Some(10), NodeState::Connected, false);
+        assert_eq!(nld.time_since_last_rx(&key), Some(Instant::now().elapsed()));
+        let last_change = Instant::now();
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.update(&key, Some(10), Some(10), NodeState::Connected, false);
+        assert_eq!(nld.time_since_last_rx(&key), Some(last_change.elapsed()));
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.update(&key, None, None, NodeState::Connected, false);
+        assert_eq!(nld.time_since_last_rx(&key), Some(last_change.elapsed()));
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.update(&key, None, None, NodeState::Connecting, false);
+        assert_eq!(nld.time_since_last_rx(&key), Some(last_change.elapsed()));
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.update(&key, None, Some(15), NodeState::Connected, false);
+        assert_eq!(nld.time_since_last_rx(&key), Some(last_change.elapsed()));
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.update(&key, Some(15), Some(15), NodeState::Connected, false);
+        assert_eq!(nld.time_since_last_rx(&key), Some(Instant::now().elapsed()));
+        let last_change = Instant::now();
+
+        time::advance(Duration::from_secs(1)).await;
+        nld.update(&key, Some(15), Some(15), NodeState::Connected, true);
+        assert_eq!(nld.time_since_last_rx(&key), Some(last_change.elapsed()));
```

Comment:
Not exactly sure what these asserts check to be honest, as I understand we advance time, insert and then check so the time of update would show to now(time of insert) but that could be only one small testcase. Could you explain why do we need so many?